### DISCUSSION TITLE
아래 코드를 보고 각 코드에서 두 counter는 독립적인지에 대해서 이유와 함께 설명해주세요.

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#85] [클로저]
- 아래 코드를 보고 각 코드에서 두 counter는 독립적인지에 대해서 이유와 함께 설명해주세요.
1.     
    ```js
    const counter = (function() {
        let cnt = 0;
        return {
            increase: function() {
                return ++cnt;
            }
        };
    })();
    
    const counter1 = counter.increase;
    const counter2 = counter.increase;
    
    console.log(counter1());
    console.log(counter1());
    
    console.log(counter2());
    console.log(counter2());
    ```

2.     
    ```js
    function makeCounter() {
      let count = 0;
    
      return function() {
        return count++;
      };
    }
    
    const counter1 = makeCounter();
    const counter2 = makeCounter();
    
    console.log(counter1())
    console.log(counter1())
    
    console.log(counter2())
    console.log(counter2())
    ```

**Comment by publdaze** [comment]
> - 아래 코드를 보고 각 코드에서 두 counter는 독립적인지에 대해서 이유와 함께 설명해주세요.
> 1.     
>     ```js
>     const counter = (function() {
>         let cnt = 0;
>         return {
>             increase: function() {
>                 return ++cnt;
>             }
>         };
>     })();
>     
>     const counter1 = counter.increase;
>     const counter2 = counter.increase;
>     
>     console.log(counter1());
>     console.log(counter1());
>     
>     console.log(counter2());
>     console.log(counter2());
>     ```
> 

- 두 counter는 의존적
- 즉시 실행 함수를 사용하여 counter 객체 생성
- 즉시 실행 함수가 호출되면 실행 컨텍스트가 생성되고, 렉시컬 환경이 바인딩되어 실행 컨텍스트 스택에 푸시됨
- 즉시 실행 함수의 실행이 끝나면 실행 컨텍스트 스택에서 실행 컨텍스트가 팝되지만
- 반환된 객체의 메소드 increase가 cnt를 참조하기 때문에 cnt는 클로저를 통해 유지
- counter1과 counter2는 반환되는 객체 프로퍼티 increase가 할당됨
- 동일한 클로저 increase 메소드를 참조하므로 cnt 식별자 또한 동일한 렉시컬 환경에서 찾을 수 있음
- 따라서 출력은 1,2,3,4가 되며, 의존적인 걸 확인할 수 있음

> 2.     
>     ```js
>     function makeCounter() {
>       let count = 0;
>     
>       return function() {
>         return count++;
>       };
>     }
>     
>     const counter1 = makeCounter();
>     const counter2 = makeCounter();
>     
>     console.log(counter1())
>     console.log(counter1())
>     
>     console.log(counter2())
>     console.log(counter2())
>     ```

- 두 counter는 독립적
- 함수 호출 시점에 실행 컨텍스트가 생성되고 그에 바인딩할 렉시컬 환경이 생성됨
- 각각의 counter에 들어갈 반환되는 중첩 함수가 참조하는 외부함수는 독립적인 렉시컬 환경을 갖게 됨
- 따라서 출력은 0,1,0,1이 됨.

**Comment by pipisebastian** [comment]
> * 아래 코드를 보고 각 코드에서 두 counter는 독립적인지에 대해서 이유와 함께 설명해주세요.
> 
> 1. ```js
>    const counter = (function() {
>        let cnt = 0;
>        return {
>            increase: function() {
>                return ++cnt;
>            }
>        };
>    })();
>    
>    const counter1 = counter.increase;
>    const counter2 = counter.increase;
>    
>    console.log(counter1());
>    console.log(counter1());
>    
>    console.log(counter2());
>    console.log(counter2());
>    ```

 -  독립적이지 않습니다. 즉시실행함수를 통해 counter를 생성하고, 내부 함수 increase 에서 상위함수(즉시실행함수)의 지역변수 cnt를 참조하므로, 클로저를 통해 외부함수의 렉시컬 환경은 계속 유지가 됩니다 . 그러나, 이는 동일한 counter 객체의 increase 메서드를 counter1, counter2 에 할당하므로, 하나의 렉시컬환경을 공유해서 사용합니다. 그렇기에 두 변수는 의존적입니다.

> 2. ```js
>    function makeCounter() {
>      let count = 0;
>    
>      return function() {
>        return count++;
>      };
>    }
>    
>    const counter1 = makeCounter();
>    const counter2 = makeCounter();
>    
>    console.log(counter1())
>    console.log(counter1())
>    
>    console.log(counter2())
>    console.log(counter2())
>    ```

- 독립적입니다. 해당 코드에서 counter1, counter2는 makeCounter 각각 호출하고, 이때, 새로운 makeCounter의 렉시컬환경이 만들어집니다. 그렇기에 독립적입니다.
